cargo-guppy/src/lib.rs

```diff
@@ -18,18 +25,136 @@ pub fn cmd_diff(json: bool, old: &str, new: &str) -> Result<(), Error> {
     Ok(())
 }
 
-pub fn cmd_count() -> Result<(), Error> {
+pub fn cmd_count() -> Result<(), anyhow::Error> {
     let lockfile = Lockfile::from_file("Cargo.lock")?;
 
     println!("Third-party Packages: {}", lockfile.third_party_packages());
 
     Ok(())
 }
 
-pub fn cmd_dups() -> Result<(), Error> {
+pub fn cmd_dups() -> Result<(), anyhow::Error> {
     let lockfile = Lockfile::from_file("Cargo.lock")?;
 
     lockfile.duplicate_packages();
 
     Ok(())
 }
+
+arg_enum! {
+    #[derive(Debug)]
+    pub enum Kind {
+        All,
+        Workspace,
+        ThirdParty,
+    }
+}
+
+struct NameVisitor;
+
+impl PackageDotVisitor for NameVisitor {
+    fn visit_package(&self, package: &PackageMetadata, mut f: DotWrite<'_, '_>) -> fmt::Result {
+        write!(f, "{}", package.name())
+    }
+
+    fn visit_link(&self, _link: DependencyLink<'_>, mut f: DotWrite<'_, '_>) -> fmt::Result {
+        write!(f, "")
+    }
+}
+
+pub fn cmd_select(
+    kind: Kind,
+    target: Option<String>,
+    output: Option<String>,
+    include_dev: bool,
+    include_build: bool,
+    omit_edges_into: &[String],
+    roots: &[String],
+) -> Result<(), anyhow::Error> {
+    let mut pkg_graph = PackageGraph::default();
+    let mut package_ids = HashSet::new();
+    let mut omitted_package_ids = HashSet::new();
+
+    let root_set: HashSet<String> = if roots.len() > 0 {
+        roots.iter().cloned().collect()
+    } else {
+        pkg_graph
+            .packages()
+            .filter_map(|meta| {
+                if meta.in_workspace() {
+                    Some(meta.name().to_string())
+                } else {
+                    None
+                }
+            })
+            .collect()
```

Comment:
This might be easier to do using https://docs.rs/guppy/0.1.2/guppy/graph/struct.PackageSelect.html#method.into_root_ids:

```
pkg_graph.select_reverse(pkg_graph.workspace().member_ids())
   .into_root_ids(DependencyDirection::Forward)
```

We should probably add an `into_root_metadatas` to mirror `into_iter_metadatas`, want to add that? :)

Reply:
Good idea. I'll add that.